06_puppy/runnerdave/map_store.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import "fmt"
+
+// MapStore represents a simple map storage for the Puppy store
+type MapStore struct {
+	puppies map[uint16]Puppy
+}
+
+// NewMapStore creates a new in-memory store with map intialised
+func NewMapStore() *MapStore {
+	var m MapStore
+	m.puppies = make(map[uint16]Puppy)
+	return &m
+}
+
+// CreatePuppy saves new puppy if not in store, if it is already returns error
+func (m *MapStore) CreatePuppy(p *Puppy) error {
+	if _, ok := m.puppies[p.ID]; ok {
+		return fmt.Errorf("puppy with id %d already exists", p.ID)
+	}
+	m.puppies[p.ID] = *p
+	return nil
+}
+
+// ReadPuppy reads store by Puppy ID
+func (m *MapStore) ReadPuppy(id uint16) (*Puppy, error) {
+	if puppy, ok := m.puppies[id]; ok {
+		return &puppy, nil
```

Comment:
I am not sure if returning a pointer is acceptable here, that may lead to a situation that some code can use the pointer to update the puppy (struct) without going through `UpdatePuppy` function, which IMO should be avoided.

Reply:
fixed, good observation